### PR TITLE
fix: retry handler

### PR DESCRIPTION
**Pull request by ronag**
Clean up and fix various issues...

### CODE REVIEW

File: lib/handler/retry-handler.js [hunk 1]
```diff
@@ -29,13 +40,13 @@ class RetryHandler {
...
-    this.dispatch = handlers.dispatch
-    this.handler = handlers.handler
-    this.opts = dispatchOpts
-    this.abort = null
-    this.aborted = false
-    this.retryOpts = {
-      retry: retryFn ?? RetryHandler[kRetryHandlerDefaultRetry],
+    this.#opts = dispatchOpts // TODO (fix): This assumes that opts are not mutated...
+    this.#dispatch = dispatch
+    this.#handler = handler
```
Comment (mcollina) [inline lib/handler/retry-handler.js:45]: I would recommend not doing this, as functions could be in there too. Is there a need?

**Comment by mcollina** [comment]
What issues did you fix? Could you add a test for them?

File: lib/handler/retry-handler.js (hunk 1 above)
Comment (ronag) [reply lib/handler/retry-handler.js:45]: Sorry, I don't understand.

**Comment by ronag** [comment]
> What issues did you fix? Could you add a test for them?

No bandwidth. Maybe in future.

**Comment by ronag** [comment]
@metcoder95 feel free to have a look if you see something you are interested in

**Comment by Uzlopak** [comment]
Can you atleast name the bugs?

File: lib/handler/retry-handler.js [hunk 2]
```diff
@@ -29,13 +40,13 @@ class RetryHandler {
...
+    this.#abort = null
+    this.#aborted = false
+    this.#retryOpts = {
+      retry: retryFn ?? defaultRetry,
```
Comment (metcoder95) [inline lib/handler/retry-handler.js:49]: This one I'm pretty sure is needed, leftover from initial ideas of it

File: lib/handler/retry-handler.js [hunk 3]
```diff
@@ -59,292 +70,227 @@ class RetryHandler {
...
-
-    // Checkpoint for resume from where we left it
-    if (this.resume != null) {
-      this.resume = null
-
-      if (statusCode !== 206) {
-        return true
+      if (this.#retryOpts.statusCodes.includes(statusCode)) {
+        // TODO (fix): Remove this path and delegate to different interceptor which will read the error body?
+        const message = `Response status code ${statusCode}${statusMessage ? `: ${statusMessage}` : ''}`
```
Comment (metcoder95) [inline lib/handler/retry-handler.js:100]: Curious about this one, what did you have in mind @ronag?

File: lib/handler/retry-handler.js [hunk 4]
```diff
@@ -59,292 +70,227 @@ class RetryHandler {
...
+        throw new ResponseStatusCodeError(message, statusCode, headers)
```
Comment (metcoder95) [inline lib/handler/retry-handler.js:101]: Why throw instead of abort?

File: lib/handler/retry-handler.js [hunk 5]
```diff
@@ -59,292 +70,227 @@ class RetryHandler {
...
+      return this.#handler.onHeaders(statusCode, rawHeaders, resume, statusMessage)
+    } else if (this.#reason == null) {
+      assert(this.#etag = null)
+      assert(this.#pos == null)
+      assert(this.#end == null)
+      assert(this.#headersSent === false)
+
+      if (headers.trailer) {
+        this.#headersSent = true
+        return this.#handler.onHeaders(statusCode, rawHeaders, resume, statusMessage)
```
Comment (metcoder95) [inline lib/handler/retry-handler.js:114]: Never thought about the `trailers`, basically if trailers are hinted we just forward do not retry?

Shall we add some flag for implementers to maybe bypass it?
I see the benefit on this

**Review by metcoder95 — COMMENTED** [review]
I'll definitely take the private properties from the PR 👍 
Following, it seems that you found a bug where the `onHeaders` were called twice? Is that correct?
If also, did you find something else and a way to reproduce it?

Thanks for the hints, I'll definitely take some improvements from here and add it maybe later this week 👍

File: lib/handler/retry-handler.js (hunk 4 above)
Comment (ronag) [reply lib/handler/retry-handler.js:101]: easier to follow the flow

File: lib/handler/retry-handler.js (hunk 3 above)
Comment (ronag) [reply lib/handler/retry-handler.js:100]: Should propagate similar to throwOnError

**Comment by ronag** [comment]
Sorry, I'm super stressed to fix this in production.

**Pull request closed without merging**